File: comfyui/ui/src/schemas/shared-validation.ts
/**
 * Shared Validation Constants and Helpers
 *
 * Common validation constraints used across all schemas.
 * Ensures consistency and prevents security issues.
 * Uses configurable limits from validationLimitsStore for power users.
 *
 * NOTE: Limits are read at schema creation time. To use updated limits,
 * schemas should be recreated or use the getter functions in refinements.
 */

import { z } from "zod";

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// Validation Constants (Fixed - Security Critical)
// ════════════════════════════════════════════════════════════════════════════

export const MAX_NAME_LENGTH = 200;
export const MAX_DESCRIPTION_LENGTH = 2000;
export const MAX_COMMENT_LENGTH = 5000;
export const MAX_TAG_LENGTH = 50;
export const MAX_TAGS_COUNT = 50;
export const MAX_PATH_LENGTH = 500;
export const MAX_ID_LENGTH = 200;
export const MAX_MIME_TYPE_LENGTH = 100;
export const MAX_FONT_FAMILY_LENGTH = 200;
export const MAX_FONT_STYLE_LENGTH = 100;
export const MAX_URL_LENGTH = 2048;
export const MAX_BASE64_LENGTH = 50 * 1024 * 1024; // 50MB max base64 data (security limit)
export const MAX_SHADER_LENGTH = 100000; // Max shader code length
export const MAX_FILENAME_LENGTH = 255;
export const MAX_ANIMATION_NAME_LENGTH = 100;
export const MAX_WARNING_LENGTH = 500;

// ════════════════════════════════════════════════════════════════════════════
// Configurable Limits (Default values - updated from store at runtime)
// ════════════════════════════════════════════════════════════════════════════

// Default values matching existing codebase constants
// These will be updated from validationLimitsStore when it loads
let _maxDimension = 8192;
let _maxFrameCount = 10000;
let _maxArrayLength = 100000;
let _maxParticles = 1000000;
let _maxLayers = 1000;
let _maxKeyframesPerProperty = 10000;
let _maxStringLength = 100000;
let _maxFPS = 120;

/**
 * Update limits (called from validationLimitsStore)
 */
export function updateLimits(limits: {
  maxDimension: number;
  maxFrameCount: number;
  maxArrayLength: number;
  maxParticles: number;
  maxLayers: number;
  maxKeyframesPerProperty: number;
  maxStringLength: number;
  maxFPS: number;
}): void {
  _maxDimension = limits.maxDimension;
  _maxFrameCount = limits.maxFrameCount;
  _maxArrayLength = limits.maxArrayLength;
  _maxParticles = limits.maxParticles;
  _maxLayers = limits.maxLayers;
  _maxKeyframesPerProperty = limits.maxKeyframesPerProperty;
  _maxStringLength = limits.maxStringLength;
  _maxFPS = limits.maxFPS;
}

/**
 * Initialize limits from store (call this after store is available)
 */
export function initializeLimits(): void {
  try {
    // Dynamic import to avoid circular dependency
    import("../stores/validationLimitsStore").then((module) => {
      const store = module.useValidationLimitsStore();
      const limits = store.getLimits();
      updateLimits({
        maxDimension: limits.maxDimension,
        maxFrameCount: limits.maxFrameCount,
        maxArrayLength: limits.maxArrayLength,
        maxParticles: limits.maxParticles,
        maxLayers: limits.maxLayers,
        maxKeyframesPerProperty: limits.maxKeyframesPerProperty,
        maxStringLength: limits.maxStringLength,
        maxFPS: limits.maxFPS,
      });
    });
  } catch (e) {
    console.warn("Failed to initialize validation limits:", e);
  }
}

/**
 * Get current max dimension
 */
export function getMaxDimension(): number {
  return _maxDimension;
}

/**
 * Get current max frame count
 */
export function getMaxFrameCount(): number {
  return _maxFrameCount;
}

/**
 * Get current max array length
 */
export function getMaxArrayLength(): number {
  return _maxArrayLength;
}

/**
 * Get current max particles
 */
export function getMaxParticles(): number {
  return _maxParticles;
}

/**
 * Get current max layers
 */
export function getMaxLayers(): number {
  return _maxLayers;
}

/**
 * Get current max keyframes per property
 */
export function getMaxKeyframesPerProperty(): number {
  return _maxKeyframesPerProperty;
}

/**
 * Get current max string length
 */
export function getMaxStringLength(): number {
  return _maxStringLength;
}

/**
 * Get current max FPS
 */
export function getMaxFPS(): number {
  return _maxFPS;
}

// Legacy exports for backward compatibility (use current values)
export const MAX_STRING_LENGTH = _maxStringLength;
export const MAX_ARRAY_LENGTH = _maxArrayLength;

// ════════════════════════════════════════════════════════════════════════════
// Helper Schema Factories
// ════════════════════════════════════════════════════════════════════════════

/** Non-empty string with max length and trimming */
export const nonEmptyString = (maxLength: number) =>
  z.string().min(1).max(maxLength).trim();

/** ISO 8601 datetime string */
export const iso8601DateTime = z.string().datetime();

/** Base64 or data URL string */
export const base64OrDataUrl = z.string().refine(
  (val) => {
    if (val.length > MAX_BASE64_LENGTH) return false;
    // Data URL format: data:[<mediatype>][;base64],<data>
    if (val.startsWith("data:")) {
      const commaIndex = val.indexOf(",");
      if (commaIndex === -1) return false;
      const header = val.substring(0, commaIndex);
      const data = val.substring(commaIndex + 1);
      return header.includes("base64") || data.length > 0;
    }
    // Base64 format: only base64 characters
    return /^[A-Za-z0-9+/=]+$/.test(val);
  },
  { message: "Must be valid base64 or data URL" }
);

/** Valid MIME type */
export const mimeType = z.string()
  .max(MAX_MIME_TYPE_LENGTH)
  .refine(
    (val) => /^[a-z][a-z0-9]*\/[a-z0-9][a-z0-9._-]*$/i.test(val),
    { message: "Must be valid MIME type (e.g., 'image/png', 'video/mp4')" }
  );

/** Hex color string (#RRGGBB or #RRGGBBAA) */
export const hexColor = z.string().regex(
  /^#([0-9A-Fa-f]{6}|[0-9A-Fa-f]{8})$/,
  { message: "Must be valid hex color (#RRGGBB or #RRGGBBAA)" }
);

/** Entity ID format - alphanumeric with underscores/hyphens */
export const entityId = z.string()
  .min(1)
  .max(MAX_ID_LENGTH)
  .regex(
    /^[a-zA-Z0-9_-]+$/,
    { message: "Must be valid ID format (alphanumeric, underscores, hyphens only)" }
  );

/** Property path (e.g., "data.text", "transform.position.x") */
export const propertyPath = z.string()
  .max(MAX_PATH_LENGTH)
  .regex(
    /^[a-zA-Z_$][a-zA-Z0-9_$.]*(\.[a-zA-Z_$][a-zA-Z0-9_$.]*)*$/,
    { message: "Must be valid property path (e.g., 'data.text', 'transform.position.x')" }
  );

/** Filename validation */
export const filename = z.string()
  .max(MAX_FILENAME_LENGTH)
  .refine(
    (val) => !/[<>:"|?*\x00-\x1F]/.test(val) && !val.endsWith(".") && !val.endsWith(" "),
    { message: "Must be valid filename (no invalid characters, no trailing dot/space)" }
  );

/** URL validation */
export const url = z.string()
  .max(MAX_URL_LENGTH)
  .url({ message: "Must be valid URL" });

/** Shader code validation */
export const shaderCode = z.string()
  .max(MAX_SHADER_LENGTH)
  .refine(
    (val) => {
      // Basic validation - check for dangerous patterns
      const dangerousPatterns = [
        /\beval\s*\(/,
        /\bFunction\s*\(/,
        /\brequire\s*\(/,
        /\bimport\s*\(/,
        /\bprocess\./,
        /\bfetch\s*\(/,
        /\bXMLHttpRequest\b/,
        /\bWebSocket\b/,
        /\bdocument\./,
        /\bwindow\./,
      ];
      return !dangerousPatterns.some((pattern) => pattern.test(val));
    },
    { message: "Shader code contains potentially dangerous patterns" }
  );

/** Array with max length (uses configurable limit) */
// Lean4/PureScript/Haskell: Explicit pattern matching - no lazy ??
export const boundedArray = <T extends z.ZodTypeAny>(schema: T, maxLength?: number) => {
  const limit = (maxLength !== null && maxLength !== undefined && typeof maxLength === "number" && Number.isFinite(maxLength) && maxLength > 0) ? maxLength : _maxArrayLength;
  return z.array(schema).max(limit);
};

/** Non-empty array */
export const nonEmptyArray = <T extends z.ZodTypeAny>(schema: T) =>
  z.array(schema).min(1);

/**
 * JSON-serializable value schema
 * Validates that a value is JSON-serializable (string, number, boolean, null, object, array)
 * Uses recursive validation to ensure nested structures are also JSON-serializable
 */
export const jsonSerializable: z.ZodTypeAny = z.lazy(() =>
  z.union([
    z.string().max(MAX_STRING_LENGTH),
    z.number().finite(),
    z.boolean(),
    z.null(),
    z.array(jsonSerializable).max(_maxArrayLength),
    z.record(z.string().max(MAX_NAME_LENGTH), jsonSerializable).refine(
      (obj) => Object.keys(obj).length <= 10000,
      { message: "Maximum 10000 keys in JSON object" }
    ),
  ])
).refine(
  (data) => {
    try {
      JSON.stringify(data);
      return true;
    } catch {
      return false;
    }
  },
  { message: "Value must be JSON-serializable" }
);

// Type export for jsonSerializable
export type JsonSerializable = z.infer<typeof jsonSerializable>;

File: comfyui-edition/src/lattice/Schema/SharedValidation.hs
-- |
-- Module      : Lattice.Schema.SharedValidation
-- Description : Shared validation functions for schema validation
-- 
-- Migrated from ui/src/schemas/shared-validation.ts
-- Provides validation functions matching Zod schema patterns
-- Used for validating data at runtime boundaries (FFI, JSON parsing)
--
-- All validation functions return Either Text a (success or error message)

{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE LambdaCase #-}

module Lattice.Schema.SharedValidation
  ( -- Validation constants
    maxNameLength
  , maxDescriptionLength
  , maxCommentLength
  , maxTagLength
  , maxTagsCount
  , maxPathLength
  , maxIdLength
  , maxMimeTypeLength
  , maxFontFamilyLength
  , maxFontStyleLength
  , maxUrlLength
  , maxBase64Length
  , maxShaderLength
  , maxFilenameLength
  , maxAnimationNameLength
  , maxWarningLength
    -- Configurable limits (with defaults)
  , ValidationLimits(..)
  , defaultValidationLimits
  , initValidationLimits
  , updateLimits
  , getLimits
  , getMaxDimension
  , getMaxDimensionPure
  , getMaxFrameCount
  , getMaxFrameCountPure
  , getMaxArrayLength
  , getMaxArrayLengthPure
  , getMaxParticles
  , getMaxParticlesPure
  , getMaxLayers
  , getMaxLayersPure
  , getMaxKeyframesPerProperty
  , getMaxKeyframesPerPropertyPure
  , getMaxStringLength
  , getMaxStringLengthPure
  , getMaxFPS
  , getMaxFPSPure
    -- Validation functions
  , validateNonEmptyString
  , validateIso8601DateTime
  , validateBase64OrDataUrl
  , validateMimeType
  , validateHexColor
  , validateEntityId
  , validatePropertyPath
  , validateFilename
  , validateUrl
  , validateShaderCode
  , validateBoundedArray
  , validateNonEmptyArray
  , validateJsonSerializable
  ) where

import Data.Text (Text)
import qualified Data.Text as T
import Data.Aeson (Value(..), encode, eitherDecode)
import qualified Data.ByteString.Lazy as BL
import Text.Regex.TDFA ((=~))
import Control.Monad (void)
import Control.Concurrent.MVar (MVar, newMVar, readMVar, swapMVar)

-- ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
--                                                   // validation // constants
-- ════════════════════════════════════════════════════════════════════════════

maxNameLength :: Int
maxNameLength = 200

maxDescriptionLength :: Int
maxDescriptionLength = 2000

maxCommentLength :: Int
maxCommentLength = 5000

maxTagLength :: Int
maxTagLength = 50

maxTagsCount :: Int
maxTagsCount = 50

maxPathLength :: Int
maxPathLength = 500

maxIdLength :: Int
maxIdLength = 200

maxMimeTypeLength :: Int
maxMimeTypeLength = 100

maxFontFamilyLength :: Int
maxFontFamilyLength = 200

maxFontStyleLength :: Int
maxFontStyleLength = 100

maxUrlLength :: Int
maxUrlLength = 2048

maxBase64Length :: Int
maxBase64Length = 50 * 1024 * 1024  -- 50MB

maxShaderLength :: Int
maxShaderLength = 100000

maxFilenameLength :: Int
maxFilenameLength = 255

maxAnimationNameLength :: Int
maxAnimationNameLength = 100

maxWarningLength :: Int
maxWarningLength = 500

-- ════════════════════════════════════════════════════════════════════════════
--                                                    // configurable // limits
-- ════════════════════════════════════════════════════════════════════════════

data ValidationLimits = ValidationLimits
  { limitsMaxDimension :: Int
  , limitsMaxFrameCount :: Int
  , limitsMaxArrayLength :: Int
  , limitsMaxParticles :: Int
  , limitsMaxLayers :: Int
  , limitsMaxKeyframesPerProperty :: Int
  , limitsMaxStringLength :: Int
  , limitsMaxFPS :: Int
  }
  deriving (Eq, Show)

defaultValidationLimits :: ValidationLimits
defaultValidationLimits = ValidationLimits
  { limitsMaxDimension = 8192
  , limitsMaxFrameCount = 10000
  , limitsMaxArrayLength = 100000
  , limitsMaxParticles = 1000000
  , limitsMaxLayers = 1000
  , limitsMaxKeyframesPerProperty = 10000
  , limitsMaxStringLength = 100000
  , limitsMaxFPS = 120
  }

-- | Initialize validation limits MVar (call once at startup)
-- Must be called before any validation functions that use limits
initValidationLimits :: IO (MVar ValidationLimits)
initValidationLimits = newMVar defaultValidationLimits

-- | Update validation limits (called from TypeScript store)
updateLimits :: MVar ValidationLimits -> ValidationLimits -> IO ()
updateLimits limits newVal = void (swapMVar limits newVal)

-- | Get current validation limits
getLimits :: MVar ValidationLimits -> IO ValidationLimits
getLimits = readMVar

-- | Get max dimension (pure version - accepts limits as parameter)
getMaxDimensionPure :: ValidationLimits -> Int
getMaxDimensionPure = limitsMaxDimension

-- | Get max dimension (IO version - reads from MVar)
getMaxDimension :: MVar ValidationLimits -> IO Int
getMaxDimension limits = limitsMaxDimension <$> readMVar limits

-- | Get max frame count (pure version)
getMaxFrameCountPure :: ValidationLimits -> Int
getMaxFrameCountPure = limitsMaxFrameCount

-- | Get max frame count (IO version)
getMaxFrameCount :: MVar ValidationLimits -> IO Int
getMaxFrameCount limits = limitsMaxFrameCount <$> readMVar limits

-- | Get max array length (pure version)
getMaxArrayLengthPure :: ValidationLimits -> Int
getMaxArrayLengthPure = limitsMaxArrayLength

-- | Get max array length (IO version)
getMaxArrayLength :: MVar ValidationLimits -> IO Int
getMaxArrayLength limits = limitsMaxArrayLength <$> readMVar limits

-- | Get max particles (pure version)
getMaxParticlesPure :: ValidationLimits -> Int
getMaxParticlesPure = limitsMaxParticles

-- | Get max particles (IO version)
getMaxParticles :: MVar ValidationLimits -> IO Int
getMaxParticles limits = limitsMaxParticles <$> readMVar limits

-- | Get max layers (pure version)
getMaxLayersPure :: ValidationLimits -> Int
getMaxLayersPure = limitsMaxLayers

-- | Get max layers (IO version)
getMaxLayers :: MVar ValidationLimits -> IO Int
getMaxLayers limits = limitsMaxLayers <$> readMVar limits

-- | Get max keyframes per property (pure version)
getMaxKeyframesPerPropertyPure :: ValidationLimits -> Int
getMaxKeyframesPerPropertyPure = limitsMaxKeyframesPerProperty

-- | Get max keyframes per property (IO version)
getMaxKeyframesPerProperty :: MVar ValidationLimits -> IO Int
getMaxKeyframesPerProperty limits = limitsMaxKeyframesPerProperty <$> readMVar limits

-- | Get max string length (pure version)
getMaxStringLengthPure :: ValidationLimits -> Int
getMaxStringLengthPure = limitsMaxStringLength

-- | Get max string length (IO version)
getMaxStringLength :: MVar ValidationLimits -> IO Int
getMaxStringLength limits = limitsMaxStringLength <$> readMVar limits

-- | Get max FPS (pure version)
getMaxFPSPure :: ValidationLimits -> Int
getMaxFPSPure = limitsMaxFPS

-- | Get max FPS (IO version)
getMaxFPS :: MVar ValidationLimits -> IO Int
getMaxFPS limits = limitsMaxFPS <$> readMVar limits

-- ════════════════════════════════════════════════════════════════════════════
--                                                   // validation // functions
-- ════════════════════════════════════════════════════════════════════════════

-- | Validate non-empty string with max length and trimming
validateNonEmptyString :: Int -> Text -> Either Text Text
validateNonEmptyString maxLen str =
  let trimmed = T.strip str
  in if T.null trimmed
    then Left "String cannot be empty"
    else if T.length trimmed > maxLen
      then Left ("String must be at most " <> T.pack (show maxLen) <> " characters")
      else Right trimmed

-- | Validate ISO 8601 datetime string
validateIso8601DateTime :: Text -> Either Text Text
validateIso8601DateTime str =
  -- Basic ISO 8601 pattern: YYYY-MM-DDTHH:mm:ss.sssZ or YYYY-MM-DDTHH:mm:ssZ
  if str =~ ("^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(\\.[0-9]{1,3})?Z?$" :: String)
    then Right str
    else Left "Must be valid ISO 8601 datetime string"

-- | Validate base64 or data URL string
validateBase64OrDataUrl :: Text -> Either Text Text
validateBase64OrDataUrl str =
  if T.length str > maxBase64Length
    then Left "Base64/data URL exceeds maximum length"
    else if T.isPrefixOf "data:" str
      then validateDataUrl str
      else validateBase64 str
  where
    validateDataUrl :: Text -> Either Text Text
    validateDataUrl urlStr =
      case T.findIndex (== ',') urlStr of
        Nothing -> Left "Data URL must contain comma separator"
        Just idx ->
          let header = T.take idx urlStr
              dataPart = T.drop (idx + 1) urlStr
          in if T.isInfixOf "base64" header || not (T.null dataPart)
            then Right urlStr
            else Left "Data URL must have base64 encoding or non-empty data"
    
    validateBase64 :: Text -> Either Text Text
    validateBase64 str' =
      if str' =~ ("^[A-Za-z0-9+/=]+$" :: String)
        then Right str'
        else Left "Must be valid base64 or data URL"

-- | Validate MIME type
validateMimeType :: Text -> Either Text Text
validateMimeType str =
  if T.length str > maxMimeTypeLength
    then Left ("MIME type must be at most " <> T.pack (show maxMimeTypeLength) <> " characters")
    else if str =~ ("^[a-z][a-z0-9]*/[a-z0-9][a-z0-9._-]*$" :: String)
      then Right str
      else Left "Must be valid MIME type (e.g., 'image/png', 'video/mp4')"

-- | Validate hex color string (#RRGGBB or #RRGGBBAA)
validateHexColor :: Text -> Either Text Text
validateHexColor str =
  if str =~ ("^#([0-9A-Fa-f]{6}|[0-9A-Fa-f]{8})$" :: String)
    then Right str
    else Left "Must be valid hex color (#RRGGBB or #RRGGBBAA)"

-- | Validate entity ID format (alphanumeric with underscores/hyphens)
validateEntityId :: Text -> Either Text Text
validateEntityId str =
  if T.null str
    then Left "Entity ID cannot be empty"
    else if T.length str > maxIdLength
      then Left ("Entity ID must be at most " <> T.pack (show maxIdLength) <> " characters")
      else if str =~ ("^[a-zA-Z0-9_-]+$" :: String)
        then Right str
        else Left "Must be valid ID format (alphanumeric, underscores, hyphens only)"

-- | Validate property path (e.g., "data.text", "transform.position.x")
validatePropertyPath :: Text -> Either Text Text
validatePropertyPath str =
  if T.length str > maxPathLength
    then Left ("Property path must be at most " <> T.pack (show maxPathLength) <> " characters")
    else if str =~ ("^[a-zA-Z_$][a-zA-Z0-9_$.]*(\\.[a-zA-Z_$][a-zA-Z0-9_$.]*)*$" :: String)
      then Right str
      else Left "Must be valid property path (e.g., 'data.text', 'transform.position.x')"

-- | Validate filename
validateFilename :: Text -> Either Text Text
validateFilename str =
  if T.length str > maxFilenameLength
    then Left ("Filename must be at most " <> T.pack (show maxFilenameLength) <> " characters")
    else if str =~ ("[<>:\"|?*\\x00-\\x1F]" :: String)
      then Left "Filename contains invalid characters"
    else if T.isSuffixOf "." str || T.isSuffixOf " " str
      then Left "Filename cannot end with dot or space"
      else Right str

-- | Validate URL
validateUrl :: Text -> Either Text Text
validateUrl str =
  if T.length str > maxUrlLength
    then Left ("URL must be at most " <> T.pack (show maxUrlLength) <> " characters")
    else if T.isPrefixOf "http://" str || T.isPrefixOf "https://" str
      then Right str
      else Left "Must be valid URL"

-- | Validate shader code (check for dangerous patterns)
validateShaderCode :: Text -> Either Text Text
validateShaderCode str =
  if T.length str > maxShaderLength
    then Left ("Shader code must be at most " <> T.pack (show maxShaderLength) <> " characters")
    else checkDangerousPatterns str
  where
    dangerousPatterns :: [String]
    dangerousPatterns =
      [ "eval\\s*\\("
      , "Function\\s*\\("
      , "require\\s*\\("
      , "import\\s*\\("
      , "process\\."
      , "fetch\\s*\\("
      , "XMLHttpRequest"
      , "WebSocket"
      , "document\\."
      , "window\\."
      ]
    
    checkDangerousPatterns :: Text -> Either Text Text
    checkDangerousPatterns code =
      if any (\pattern -> code =~ pattern) dangerousPatterns
        then Left "Shader code contains potentially dangerous patterns"
        else Right code

-- | Validate bounded array (with max length)
validateBoundedArray :: Int -> [a] -> Either Text [a]
validateBoundedArray maxLen arr =
  if length arr > maxLen
    then Left ("Array must have at most " <> T.pack (show maxLen) <> " items")
    else Right arr

-- | Validate non-empty array
validateNonEmptyArray :: [a] -> Either Text [a]
validateNonEmptyArray arr =
  if null arr
    then Left "Array cannot be empty"
    else Right arr

-- | Validate JSON-serializable value
validateJsonSerializable :: Value -> Either Text Value
validateJsonSerializable val =
  case eitherDecode (encode val) :: Either String Value of
    Left err -> Left ("Value must be JSON-serializable: " <> T.pack err)
    Right _ -> Right val
